v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerEndToEndIT.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.beam.it.gcp.spanner.matchers.SpannerAsserts;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * An integration test for {@link DataStreamToSpanner} Flex template which tests use-cases where a
+ * session file is required.
+ */
+@Category({TemplateIntegrationTest.class, SkipDirectRunnerTest.class})
+@TemplateIntegrationTest(DataStreamToSpanner.class)
+@RunWith(JUnit4.class)
+public class DataStreamToSpannerEndToEndIT extends DataStreamToSpannerITBase {
+
+  private static final String TABLE1 = "Authors";
+  private static final String TABLE2 = "Books";
+  private static final String TABLE3 = "Genre";
+  private static PipelineLauncher.LaunchInfo jobInfo;
+  private static HashSet<DataStreamToSpannerEndToEndIT> testInstances = new HashSet<>();
+  public static PubsubResourceManager pubsubResourceManager;
+  public static SpannerResourceManager spannerResourceManager;
+  private static final String SPANNER_DDL_RESOURCE =
+      "DataStreamToSpannerEndToEndIT/spanner-schema.sql";
+  private static final String SESSION_FILE_RESOURCE =
+      "DataStreamToSpannerEndToEndIT/mysql-session.json";
+
+  /**
+   * Setup resource managers and Launch dataflow job once during the execution of this test class.
+   */
+  @Before
+  public void setUp() throws IOException {
+    // Prevent cleaning up of dataflow job after a test method is executed.
+    skipBaseCleanup = true;
+    synchronized (DataStreamToSpannerEndToEndIT.class) {
+      testInstances.add(this);
+      if (jobInfo == null) {
+        spannerResourceManager = setUpSpannerResourceManager();
+        pubsubResourceManager = setUpPubSubResourceManager();
+        createSpannerDDL(spannerResourceManager, SPANNER_DDL_RESOURCE);
+        jobInfo =
+            launchDataflowJob(
+                getClass().getSimpleName(),
+                SESSION_FILE_RESOURCE,
+                null,
+                "EndToEndIT",
+                spannerResourceManager,
+                pubsubResourceManager,
+                new HashMap<>() {
+                  {
+                    put("inputFileFormat", "avro");
+                  }
+                },
+                null,
+                null);
+      }
+    }
+  }
+
+  /** Cleanup dataflow job and all the resources and resource managers. */
+  @AfterClass
+  public static void cleanUp() throws IOException {
+    for (DataStreamToSpannerEndToEndIT instance : testInstances) {
+      instance.tearDownBase();
+    }
+    ResourceManagerUtils.cleanResources(spannerResourceManager, pubsubResourceManager);
+  }
+
+  @Test
+  public void endToEndMigrationTest() {
+    // Construct a ChainedConditionCheck with 2 stages.
+    // 1. Send initial wave of events
+    // 2. Wait on Spanner to have events
+    ChainedConditionCheck conditionCheck =
```

Comment:
We want to simulate DLQ retry processing by inserting child row first and then insert corresponding parent later, hence the mysql-schema should not contain FK relationship between Authors and Books.

The test case should look like this:
1. upload Books avro file containing few rows referring author 1 and few rows referring author 2.
2. upload genre avro file
3. upload author avro file containing author 1.
4. check for Author table to contain 1 row. This ensures that current wave of avro files are almost processed by the template.

Add a small sleep after Line No. 135

1. Then upload avro file containing Author number 2 (and others)
2. Wait for all rows to appear in author table, books table and genre table.
3. Then assert table contents.

Maybe this is what you are doing, it will be clear if you add the SQL statements used to create the avro files in the resources folder. Also, add comments here in this test to help others understand the test case.

Reply:
Hi Darshan. So this is what I have done. Uploaded child rows(Books), one parent row(Authors) on which none of the child rows depends and the genre table. Now since only one of these will get migrated, we check if 1 row is migrated to Authors. In this duration the Child rows(Books) have gone to dlq for retry. Then we upload the three parents rows(Author) on which the previous 3 child rows depend. We now check that both child(books) and parents(author) rows are migrated

Reply:
I will add the SQLs